Content collections build: "Vite module runner has been closed."

### Astro Info

```block
Astro                    v5.0.3
Node                     v23.2.0
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             none
```

### Describe the Bug

I read files to generate a content collection. Works fine during development, but fails during build with this output:

```
❯ astro build
17:09:47 [content] Syncing content
17:09:47 [content] Content config changed
17:09:47 [content] Clearing content store
Vite module runner has been closed.
  Location:
    /home/projects/withastro-astro-7t9i3ppv/node_modules/vite/dist/node/module-runner.js:1217:13
  Stack trace:
    at SSRCompatModuleRunner.getModuleInformation (file:///home/projects/withastro-astro-7t9i3ppv/node_modules/vite/dist/node/module-runner.js:1217:13)
    at request (file:///home/projects/withastro-astro-7t9i3ppv/node_modules/vite/dist/node/module-runner.js:1243:99)
    at ./examples/csv/grammar.txt (/home/projects/withastro-astro-7t9i3ppv/src/content.config.ts:7:89)
    at eval (file:///home/projects/withastro-astro-7t9i3ppv/node_modules/zod/lib/index.mjs:3492:40)
    at eval (file:///home/projects/withastro-astro-7t9i3ppv/node_modules/astro/dist/content/content-layer.js:188:18)
```

### What's the expected result?

Should behave identically to dev.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-7t9i3ppv?file=src%2Fcontent.config.ts

In the console, run `astro build`.

### Participation

- [x] I am willing to submit a pull request for this issue.

I ran into this issue in my website, too.
I started a new blog framework with latest AstroWind template with astro@5.0.3.

When I was using [notion-astro-loader](https://github.com/NotWoods/travel/tree/main/packages/notion-astro-loader) to build my content layer collection with notion, the same error happened.

I looked up the loader's repo and find the related code. It just uses the `getImage` API.

```typescript
#fetchImage = async (imageFileObject) => {
        try {
            const fetchedImageData = await fileToImageAsset(imageFileObject);
            this.#imagePaths.push(fetchedImageData.src);
            return fetchedImageData.src;
        }
        catch (error) {
            console.error(error)
            this.#logger.error(`Failed to fetch image when rendering page.
Have you added \`image: { remotePatterns: [{ protocol: "https", hostname: "*.amazonaws.com" }] }\` to your Astro config file?\n
Error: ${getErrorMessage(error)}`);
            // Fall back to using the remote URL directly.
            return fileToUrl(imageFileObject);
        }
    };

/**
 * Extract and locally cache the image from a file object.
 * @see https://developers.notion.com/reference/file-object
 */
export async function fileToImageAsset(
  file: FileObject,
): Promise<GetImageResult> {
  return getImage({
    src: fileToUrl(file),
    inferSize: true,
  });
}

export function fileToUrl(file: FileObject): string;
export function fileToUrl(file: FileObject | null): string | undefined;
export function fileToUrl(file: FileObject | null): string | undefined {
  switch (file?.type) {
    case "external":
      return file.external.url;
    case "file":
      return file.file.url;
    default:
      return undefined;
  }
}

```

Here is the stack trace. It seems that it is caused by dynamic import, which is mentioned at a closed Vite issue.

```
20:36:43 [ERROR] [page 02f9d03e-aa37-4ad3-b889-d6b4c20bfcb7 (Name 前端开发要知道的网络基础)] Failed to fetch image when rendering page.
Have you added `image: { remotePatterns: [{ protocol: "https", hostname: "*.amazonaws.com" }] }` to your Astro config file?

Error: There was an error loading the configured image service. Please see the stack trace for more information.
InvalidImageService: There was an error loading the configured image service. Please see the stack trace for more information.
    at eval (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_rollup@4.28.1_terser@5.36.0_typescript@5.7.2_yaml@2.5.0/node_modules/astro/dist/assets/internal.js:25:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getConfiguredImageService (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_rollup@4.28.1_terser@5.36.0_typescript@5.7.2_yaml@2.5.0/node_modules/astro/dist/assets/internal.js:21:34)
    at async getImage (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_rollup@4.28.1_terser@5.36.0_typescript@5.7.2_yaml@2.5.0/node_modules/astro/dist/assets/internal.js:55:19)
    at async getImage (astro:assets:32:42)
    at async #fetchImage (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:187:38)
    at async listBlocks (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:79:25)
    at async awaitAll (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:54:22)
    at async NotionPageRenderer.render (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:163:28)
    at async Promise.all (index 19) {
  loc: undefined,
  title: 'Error while loading image service.',
  hint: undefined,
  frame: undefined,
  type: 'AstroError',
  cause: Error: Vite module runner has been closed.
      at SSRCompatModuleRunner.getModuleInformation (file:///Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/vite@6.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_terser@5.36.0_yaml@2.5.0/node_modules/vite/dist/node/module-runner.js:1196:13)
      at SSRCompatModuleRunner.cachedModule (file:///Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/vite@6.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_terser@5.36.0_yaml@2.5.0/node_modules/vite/dist/node/module-runner.js:1186:21)
      at request (file:///Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/vite@6.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_terser@5.36.0_yaml@2.5.0/node_modules/vite/dist/node/module-runner.js:1222:99)
      at dynamicRequest (file:///Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/vite@6.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_terser@5.36.0_yaml@2.5.0/node_modules/vite/dist/node/module-runner.js:1224:124)
      at getConfiguredImageService (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_rollup@4.28.1_terser@5.36.0_typescript@5.7.2_yaml@2.5.0/node_modules/astro/dist/assets/internal.js:21:40)
      at getImage (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_rollup@4.28.1_terser@5.36.0_typescript@5.7.2_yaml@2.5.0/node_modules/astro/dist/assets/internal.js:55:25)
      at getImage (astro:assets:32:82)
      at fileToImageAsset (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/format.js:33:46)
      at #fetchImage (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:187:86)
      at listBlocks (/Users/bytedance/Code/personal/notion-astro-rev/node_modules/.pnpm/notion-astro-loader@0.4.0_astro@5.0.3_@types+node@22.5.0_jiti@1.21.6_lightningcss@1.27.0_roll_2ibjhaoyvb5densitsoeosfrn4/node_modules/notion-astro-loader/dist/render.js:79:31)
}

```

--- a/.changeset/fix-vite-runner-closed.md
+++ b/.changeset/fix-vite-runner-closed.md
@@ -1,0 +1,7 @@
+---
+'astro': patch
+---
+
+Fixes content layer loaders that use dynamic imports
+
+Content collection loaders can now use `await import()` and `import.meta.glob()` to dynamically import modules during build. Previously, these would fail with "Vite module runner has been closed."

--- a/packages/astro/src/core/sync/index.ts
+++ b/packages/astro/src/core/sync/index.ts
@@ -3,7 +3,7 @@ import { dirname, relative } from 'node:path';
 import { performance } from 'node:perf_hooks';
 import { fileURLToPath } from 'node:url';
 import colors from 'piccolore';
-import { createServer, type FSWatcher, type HotPayload } from 'vite';
+import { createServer, type FSWatcher, type HotPayload, type ViteDevServer } from 'vite';
 import { syncFonts } from '../../assets/fonts/sync.js';
 import { CONTENT_TYPES_FILE } from '../../content/consts.js';
 import { getDataStoreFile, globalContentLayer } from '../../content/content-layer.js';
@@ -127,36 +127,45 @@ export async function syncInternal({
 	const timerStart = performance.now();
 
 	if (!skip?.content) {
-		await syncContentCollections(settings, { mode, fs, logger });
-		settings.timer.start('Sync content layer');
+		// Create the Vite server once and keep it alive for both content config loading
+		// and content layer sync. This is needed because loaders may use dynamic imports
+		// which require the Vite server to be running. See https://github.com/withastro/astro/issues/12689
+		const tempViteServer = await createTempViteServer(settings, { mode, fs, logger });
 
-		let store: MutableDataStore | undefined;
 		try {
-			const dataStoreFile = getDataStoreFile(settings, isDev);
-			store = await MutableDataStore.fromFile(dataStoreFile);
-		} catch (err: any) {
-			logger.error('content', err.message);
-		}
-		if (!store) {
-			logger.error('content', 'Failed to load content store');
-			return;
-		}
+			await syncContentCollections(settings, { fs, logger, viteServer: tempViteServer });
+			settings.timer.start('Sync content layer');
 
-		const contentLayer = globalContentLayer.init({
-			settings,
-			logger,
-			store,
-			watcher,
-		});
-		if (watcher) {
-			contentLayer.watchContentConfig();
+			let store: MutableDataStore | undefined;
+			try {
+				const dataStoreFile = getDataStoreFile(settings, isDev);
+				store = await MutableDataStore.fromFile(dataStoreFile);
+			} catch (err: any) {
+				logger.error('content', err.message);
+			}
+			if (!store) {
+				logger.error('content', 'Failed to load content store');
+				return;
+			}
+
+			const contentLayer = globalContentLayer.init({
+				settings,
+				logger,
+				store,
+				watcher,
+			});
+			if (watcher) {
+				contentLayer.watchContentConfig();
+			}
+			await contentLayer.sync();
+			if (!skip?.cleanup) {
+				// Free up memory (usually in builds since we only need to use this once)
+				contentLayer.dispose();
+			}
+			settings.timer.end('Sync content layer');
+		} finally {
+			await tempViteServer.close();
 		}
-		await contentLayer.sync();
-		if (!skip?.cleanup) {
-			// Free up memory (usually in builds since we only need to use this once)
-			contentLayer.dispose();
-		}
-		settings.timer.end('Sync content layer');
 	} else {
 		const paths = getContentPaths(
 			settings.config,
@@ -208,23 +217,13 @@ function writeInjectedTypes(settings: AstroSettings, fs: typeof fsMod) {
 }
 
 /**
- * Generate content collection types, and then returns the process exit signal.
- *
- * A non-zero process signal is emitted in case there's an error while generating content collection types.
- *
- * This should only be used when the callee already has an `AstroSetting`, otherwise use `sync()` instead.
- * @internal
- *
- * @param {SyncOptions} options
- * @param {AstroSettings} settings Astro settings
- * @param {typeof fsMod} options.fs The file system
- * @param {LogOptions} options.logging Logging options
- * @return {Promise<ProcessExit>}
+ * Creates a temporary Vite server for use during content sync operations.
+ * This server is needed to load the content config and for loaders that use dynamic imports.
  */
-async function syncContentCollections(
+async function createTempViteServer(
 	settings: AstroSettings,
 	{ mode, logger, fs }: Required<Pick<SyncOptions, 'mode' | 'logger' | 'fs'>>,
-): Promise<void> {
+): Promise<ViteDevServer> {
 	const routesList = await createRoutesList(
 		{
 			settings,
@@ -234,7 +233,6 @@ async function syncContentCollections(
 		{ dev: true, skipBuildOutputAssignment: true },
 	);
 
-	// Needed to load content config
 	const tempViteServer = await createServer(
 		await createVite(
 			{
@@ -274,13 +272,34 @@ async function syncContentCollections(
 		return hotSend(payload);
 	};
 
+	return tempViteServer;
+}
+
+/**
+ * Generate content collection types, and then returns the process exit signal.
+ *
+ * A non-zero process signal is emitted in case there's an error while generating content collection types.
+ *
+ * This should only be used when the callee already has an `AstroSetting`, otherwise use `sync()` instead.
+ * @internal
+ *
+ * @param {SyncOptions} options
+ * @param {AstroSettings} settings Astro settings
+ * @param {typeof fsMod} options.fs The file system
+ * @param {LogOptions} options.logging Logging options
+ * @return {Promise<ProcessExit>}
+ */
+async function syncContentCollections(
+	settings: AstroSettings,
+	{ logger, fs, viteServer }: { logger: Logger; fs: typeof fsMod; viteServer: ViteDevServer },
+): Promise<void> {
 	try {
 		const contentTypesGenerator = await createContentTypesGenerator({
 			contentConfigObserver: globalContentConfigObserver,
 			logger: logger,
 			fs,
 			settings,
-			viteServer: tempViteServer,
+			viteServer,
 		});
 		await contentTypesGenerator.init();
 
@@ -322,7 +341,5 @@ async function syncContentCollections(
 			},
 			{ cause: e },
 		);
-	} finally {
-		await tempViteServer.close();
 	}
 }

--- a/packages/astro/test/content-layer.test.js
+++ b/packages/astro/test/content-layer.test.js
@@ -278,6 +278,16 @@ describe('Content Layer', () => {
 			});
 		});
 
+		// Regression test for https://github.com/withastro/astro/issues/12689
+		it('returns a collection from a loader that uses dynamic import', async () => {
+			assert.ok(json.hasOwnProperty('dynamicImportLoader'));
+			assert.ok(Array.isArray(json.dynamicImportLoader));
+			// Should have loaded cats.json via dynamic import
+			const ids = json.dynamicImportLoader.map((item) => item.data.id);
+			assert.ok(ids.includes('siamese'), 'Should include siamese cat');
+			assert.ok(ids.includes('tabby'), 'Should include tabby cat');
+		});
+
 		it('transforms a reference id to a reference object', async () => {
 			assert.ok(json.hasOwnProperty('entryWithReference'));
 			assert.deepEqual(json.entryWithReference.data.cat, { collection: 'cats', id: 'tabby' });

--- a/packages/astro/test/fixtures/content-layer/src/content.config.ts
+++ b/packages/astro/test/fixtures/content-layer/src/content.config.ts
@@ -365,6 +365,23 @@ const renderMarkdownTest = defineCollection({
 	}),
 });
 
+// Collection that uses dynamic import in loader - tests fix for #12689
+const dynamicImport = defineCollection({
+	loader: async () => {
+		// This dynamic import would fail with "Vite module runner has been closed" before the fix
+		const data = await import('./data/cats.json');
+		return data.default;
+	},
+	schema: z.object({
+		breed: z.string(),
+		id: z.string(),
+		size: z.string(),
+		origin: z.string(),
+		lifespan: z.string(),
+		temperament: z.array(z.string()),
+	}),
+});
+
 export const collections = {
 	blog,
 	dogs,
@@ -386,6 +403,7 @@ export const collections = {
 	rodents,
 	rockets,
 	renderMarkdownTest,
+	dynamicImport,
 	notADirectory,
 	nothingMatches
 };

--- a/packages/astro/test/fixtures/content-layer/src/pages/collections.json.js
+++ b/packages/astro/test/fixtures/content-layer/src/pages/collections.json.js
@@ -48,6 +48,9 @@ export async function GET() {
 
 	const renderMarkdownTest = await getEntry('renderMarkdownTest', 'with-frontmatter');
 	const renderMarkdownWithImage = await getEntry('renderMarkdownTest', 'with-image');
+
+	// Test for #12689 - dynamic import in loader
+	const dynamicImportLoader = await getCollection('dynamicImport');
 	
 	return new Response(
 		devalue.stringify({
@@ -77,6 +80,7 @@ export async function GET() {
 			spacecraftNoBody: spacecraftNoBody.map(({id, body}) => ({id, body})),
 			renderMarkdownTest,
 			renderMarkdownWithImage,
+			dynamicImportLoader,
 		})
 	);
 }

